### PR TITLE
[bitcode] Implement generic and wrapper deduplication

### PR DESCRIPTION
**Pull request by alexanderkyte**
Here we use the COMDAT/linkonce linker features to deduplicate methods. When two wrappers or generic instantiations with the same name are linked together statically, the linker will remove the extra versions of method bodies.

### CODE REVIEW

**Comment by alexanderkyte** [comment]
aottest = BCL + mono + nunitlite + corlib test suite statically linked

Command: `size -A aottest`

Default:
```
aottest  :
section                   size        addr
.interp                     28     4194928
.note.ABI-tag               32     4194956
.note.gnu.build-id          36     4194988
.gnu.hash                  360     4195024
.dynsym                  11424     4195384
.dynstr                   7039     4206808
.gnu.version               952     4213848
.gnu.version_r             528     4214800
.rela.dyn                 1680     4215328
.rela.plt                10224     4217008
.init                       26     4227232
.plt                      6832     4227264
.text                 84569551     4234096
.fini                        9    88803648
.rodata               12173824    88803680
.eh_frame_hdr          1053476   100977504
.eh_frame              3917684   102030984
.gcc_except_table       510148   105948668
.tbss                       72   108558576
.init_array                784   108558576
.fini_array                  8   108559360
.jcr                         8   108559368
.data.rel.ro           1126304   108559392
.dynamic                   592   109685696
.got                       488   109686288
.got.plt                  3432   109686784
.data                    19160   109690240
.bss                   1375488   109709408
.comment                   262           0
Total                104790451
```

Stripped, linkonce:
```
aottest  :
section                   size        addr
.interp                     28     4194928
.note.ABI-tag               32     4194956
.note.gnu.build-id          36     4194988
.gnu.hash                  360     4195024
.dynsym                  11424     4195384
.dynstr                   7039     4206808
.gnu.version               952     4213848
.gnu.version_r             528     4214800
.rela.dyn                 1680     4215328
.rela.plt                10224     4217008
.init                       26     4227232
.plt                      6832     4227264
.text                 84569247     4234096
.fini                        9    88803344
.rodata               12173856    88803360
.eh_frame_hdr          1053460   100977216
.eh_frame              3917620   102030680
.gcc_except_table       510148   105948300
.tbss                       72   108558592
.init_array                784   108558592
.fini_array                  8   108559376
.jcr                         8   108559384
.data.rel.ro           1126304   108559392
.dynamic                   592   109685696
.got                       488   109686288
.got.plt                  3432   109686784
.data                    19160   109690240
.bss                   1375456   109709408
.comment                   262           0
Total                104790067 
```

File: mono/mini/mini-runtime.c [hunk 1]
```diff
@@ -1517,8 +1517,9 @@ mono_resolve_patch_target (MonoMethod *method, MonoDomain *domain, guint8 *code,
 			if (run_cctors) {
 				target = mono_lookup_pinvoke_call (patch_info->data.method, &exc_class, &exc_arg);
 				if (!target) {
+					fprintf (stderr, "Unable to resolve pinvoke method '%s' Re-run with MONO_LOG_LEVEL=debug for more information.\n", mono_method_full_name (patch_info->data.method, TRUE));
 					if (mono_aot_only) {
-						mono_error_set_exception_instance (error, mono_exception_from_name_msg (mono_defaults.corlib, "System", exc_class, exc_arg));
+						/*mono_error_set_exception_instance (error, mono_exception_from_name_msg (mono_defaults.corlib, "System", exc_class, exc_arg));*/
```
Comment (xmcclure) [inline mono/mini/mini-runtime.c]: Why comment this out instead of just removing it?

File: mono/utils/mono-dl-posix.c [hunk 2]
```diff
@@ -38,6 +38,7 @@ mono_dl_get_so_suffixes (void)
 {
 	static const char *suffixes[] = {
 		".so",
+		".so.6",
```
Comment (xmcclure) [inline mono/utils/mono-dl-posix.c]: I would love to see a comment explaining why this change was made.

File: mono/mini/mini-runtime.c [hunk 3]
```diff
@@ -137,7 +137,7 @@ static GSList *tramp_infos;
 
 static void register_icalls (void);
 
-static gboolean mini_profiler_enabled (void) { return mini_enable_profiler; }
+static gboolean mini_profiler_enabled (void) { return FALSE; }
```
Comment (xmcclure) [inline mono/mini/mini-runtime.c]: What is the impact of this change?

File: mono/mini/mini-llvm.c [hunk 4]
```diff
@@ -8879,7 +9023,7 @@ mono_llvm_emit_aot_module (const char *filename, const char *cu_name)
 		}
 	}
 
-#if 1
+#if 0
```
Comment (xmcclure) [inline mono/mini/mini-llvm.c]: Why is this change made? What is its impact?

File: mono/mini/aot-compiler.c [hunk 5]
```diff
@@ -8059,48 +8059,370 @@ append_mangled_signature (GString *s, MonoMethodSignature *sig)
 	return TRUE;
 }
 
-/*
```
Comment (vargaz) [inline mono/mini/aot-compiler.c]: This looks ok.

File: mono/metadata/debug-helpers.c [hunk 6]
```diff
@@ -280,8 +280,8 @@ mono_signature_full_name (MonoMethodSignature *sig)
 	return result;
 }
 
-static void
```
Comment (vargaz) [inline mono/metadata/debug-helpers.c]: Looks ok.

File: mono/mini/aot-compiler.c [hunk 7]
```diff
@@ -8215,6 +8215,52 @@ append_mangled_wrapper_subtype (GString *s, WrapperSubtype subtype)
 	g_string_append_printf (s, "%s_", label);
 }
```
Comment (vargaz) [inline mono/mini/aot-compiler.c]: Looks ok.

File: mono/mini/aot-compiler.c [hunk 8]
```diff
@@ -8223,6 +8223,9 @@ sanitize_mangled_string (const char *input)
 	for (int i=0; input [i] != '\0'; i++) {
```
Comment (vargaz) [inline mono/mini/aot-compiler.c]: Looks ok.

**Comment by vargaz** [comment]
The AOT name mangling changes look ok.
